### DISCUSSION TITLE
브라우저 저장소에 대해서 설명해 주세요

### DISCUSSION THREAD

**Discussion by javascriptbiggosoo** [wanted-9th-3team/frontend-study#9] [프론트 기술 질문 공간] [answered]
ex) 로컬 스토리지, 세션 스토리지, 쿠키

**Comment by michoball** [comment]
HTML5 부터 들어오게 된 웹에 key value 형태로 데이터를 저장할 수 있는 저장소 입니다. 

session storage & local storage, cookies, 그리고 indexed DB 정도가 있습니다. 

session 과 local storage  는 key/ value의 형태로 값을 저장할 수 있고 세션 id 같은 간단한 데이터를 저장할 때 사용합니다. 그리고 유저나 JS로 접근이 가능합니다. 

cookie도 비슷한데 서버 쪽에서만 Js로 접근이 가능합니다. 가져오는 데이터는 string으로 저장됩니다.  서버와 data를 주고 받고 data 만료시간을 정할 수 있기때문에 많이 사용됩니다.

앞서 말씀드린 cookie와 session&local storage 와 달리 많은 양의 데이터를 저장하고 싶을 때 사용하는 것이 indexed DB 입니다.  위 저장소에 데이터를 JSON.stringify() 할 필요없이 객체 형태로 저장이 가능하다고 알고 있습니다.

**Comment by WooGie911** [comment]
대표적으로 쿠키와 세션으로 구분 할 수 있습니다.

먼저 쿠키란 사용자의 브라우저에 저장되고, 통신할 때 HTTP 헤더에 포함되는 텍스트 데이터 파일입니다. 
클라이언트에 300개 까지 쿠키 저장이 가능하고 하나의 도메인당 20개의 값만 가질 수 있습니다. 
특정 웹사이트에 접속할 때 발생하며 사용자의 인증을 도와주고 만료기간이 있어 시간이 지나면 자동으로 삭제됩니다. 

그리고 세션이란 일정 시간 동안 같은 사용자(브라우저)로부터 들어오는 일련의 요구를 하나의 상태로 보고, 그 상태를 유지시키는 기술입니다. 방문자가 웹 서버에 접속해 있는 상태를 하나의 단위로 보고 그것을 세션이라고 합니다. 
브라우저를 닫거나, 서버에서 세션을 삭제했을 때만 삭제되므로 쿠키보다 비교적 보안이 좋습니다.

**Reply by michoball** [reply]
브라우저 저장소를 얘기할 때 세션은 세션 스토리지가 아닐까요? 
쿠키와 세션에서 세션은 서버에 저장되는 값이니까 브라우저 저장소라고 할 수 없지 않을까요?

**Reply by WooGie911** [reply]
맞는 말씀 같습니다.
제가 질문을 제대로 숙지하지 못헀네요.
브라우저 저장소를 이야기 한다면 쿠키와 웹 스토리지로 볼 수 있을 것 같습니다.

웹 스토리지(Web Storage)는 클라이언트에 데이터를 저장할 수 있도록 HTML5부터 추가된 저장소입니다.
간단한 Key-Value 스토리지 형태이며, 쿠키와 달리 자동 전송의 위험성이 없고, 
오리진(Origin)(도메인,프로토콜,포트) 단위로 접근이 제한되는 특성 덕분에 CSRF로 부터 안전합니다. 
또한 쿠키보다 큰 저쟝 용량 지원(모바일 2.5MB, 데스크탑 5~10MB)하며, 
서버가 HTTP 헤더를 통해 스토리지 객체를 조작할 수 없습니다.(웹 스토리지 객체 조작은 JavaScript 내에서만 수행합니다.) 
또한 오직 문자형(string) 데이터 타입만 지원합니다. 
로컬 스토리지(Local Storage)와 세션 스토리지(Session Storage)가 있으며, 같은 Storage 객체를 상속하기 때문에 메서드가 동일합니다.

로컬 스토리지(localStorage)
사용자가 데이터를 지우지 않는 이상, 브라우저나 OS를 종료해도 계속 브라우저에 남아있습니다. (영구성) 
단, 동일한 브라우저를 사용할 때만 해당합니다. 
주로 지속적으로 필요한 데이터 저장 할 때 사용합니다. (자동 로그인 등)

세션 스토리지(sessionStorage)
데이터가 오리진 뿐만 아니라 브라우저 탭에도 종속되기 때문에, 윈도우나 브라우저 탭을 닫을 경우 제거됩니다.
일시적으로 필요한 데이터 저장 할 때 사용합니다. (일회성 로그인 정보, 입력폼 저장 등)

**Comment by skyhanull** [comment]
대표적으로 스토리지와 쿠키로 나뉩니다
쿠키는 4kb를 가진 파일로써 서버에 요청을 보낼때나 응답을 받을 때 헤더안에 넣어져서 옵니다
또한 만료시간을 정할 수 있습니다

스토리지는 세션과 로컬 스토리지로 나뉩니다. 스토리지 형태는 key와 value로 이루어져있습니다.
세션스토리지는 브라우저 창이 닫히면 같이 사라집니다
로컬스토리지는 브라우저 창을 닫아도 사라지지 않는다는 차이가 있습니다

매번 헤더에 담겨져 가는 쿠키보다 스토리지가 비교적 보안이 좋습니다

**Comment by javascriptbiggosoo** [comment, marked as answer]
### [After Discussion]

브라우저 저장소에는 대표적으로 쿠키, 로컬스토리지, 세션 스토리지 (그리고 indexed DB) 가 있습니다. 
( 이 중 indexedDB 에 대해서는 클라이언트 사이드의 복잡한 데이터를 저장할 수 있고 다른 저장소에 비해 많은 양의 데이터를 저장할 수 있다는 것 이외에는 잘은 몰라서 )  → 빼도 될 듯 설명하실 분만 넣기
다른 종류의 저장소에 대해서 설명을 드리겠습니다.  

우선 쿠키는 4kb 정도의 저장소이며 서버가 요청 응답 시 http header에 실려 보내집니다.

주로 로그인 정보, 사용자 설정, 방문한 페이지 등의 정보를 저장, 브라우저가 종료되어도 남아있을 수 있으며, 만료 날짜를 설정하여 일정 시간 이후에 자동으로 삭제할 수 있습니다. 

웹 스토리지는 클라이언트에 데이터를 저장할 수 있도록 HTML5부터 추가된 저장소입니다.

간단한 Key-Value 스토리지 형태이며, 쿠키와 달리 자동 전송의 위험성이 없고, 오리진(Origin)(도메인,프로토콜,포트) 단위로 접근이 ( 동일 출처 내에서만 ) 제한되는 특성 덕분에 보안( CSRF )에 좀 더 안전합니다.

스토리지는 브라우저에 데이터를 저장하는 방식으로, 쿠키와 다르게 더 많은 데이터를 저장할 수 있습니다.

이 웹 스토리지에는 로컬 스토리지와 세션 스토리지가 있습니다.

로컬 스토리지는 브라우저에 데이터를 저장하여, 브라우저를 종료해도 정보가 남아있을 수 있습니다.

세션 스토리지는 로컬 스토리지와 다르게 브라우저를 종료하면 데이터가 삭제됩니다.
